runtime-sdk/src/dispatcher.rs

```diff
@@ -183,6 +199,65 @@ impl<R: Runtime> Dispatcher<R> {
         })
     }
 
+    fn dispatch_message(
+        &self,
+        mut ctx: &mut DispatchContext<'_>,
+        handler_name: String,
+        message_event: MessageEvent,
+        handler_ctx: cbor::Value,
+    ) -> Result<(), modules::core::Error> {
+        // Perform message handler lookup.
+        let method_info = match self
+            .consensus_message_handlers
+            .lookup_handler(&handler_name)
+        {
+            Some(method_info) => method_info,
+            None => {
+                // Message handler not found.
+                return Err(modules::core::Error::InvalidMethod);
+            }
+        };
+
+        (method_info.handler)(&method_info, &mut ctx, message_event, handler_ctx);
+
+        Ok(())
+    }
+
+    fn handle_last_round_messages(&self, ctx: &mut DispatchContext<'_>) {
+        let message_events = ctx.runtime_round_results().messages.clone();
+
+        let store =
+            storage::TypedStore::new(storage::PrefixStore::new(ctx.runtime_state(), &MODULE));
+        let mut handlers: BTreeMap<u32, types::message::MessageEventHookInvocation> =
+            store.get(state::HANDLERS).unwrap_or_default();
+
+        for event in message_events {
+            let handler = handlers
+                .remove(&event.index)
+                .expect("message event hook should exist");
+            self.dispatch_message(ctx, handler.hook_name, event, handler.payload)
+                .expect("message event should be handled");
+        }
+        if !handlers.is_empty() {
+            panic!("unhandled message event")
```

Comment:
Instead of panicking this could also propagate the error as a `RuntimError`? And you could use the logger from `oasis-core-runtime` (BTW, should we add a logger helper to `ctx` to be able to easily get the logger from anywhere?) to report which handlers remained.

Reply:
Done